mhkit/wave/contours.py

```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+            "'PCA' method.")
+
+    pca_fit = _principal_component_analysis(x1, x2, bin_size)
+
+    # Calculate line where Hs = 0 to avoid sampling Hs in negative space
+    t_zeroline = np.linspace(2.5, 30, 1000)
+    t_zeroline = np.transpose(t_zeroline)
+    h_zeroline = np.zeros(len(t_zeroline))
+
+    # Transform zero line into principal component space
+    coeff = pca_fit['principal_axes']
+    shift = pca_fit['shift']
+    comp_zeroline = np.dot(np.transpose(np.vstack([h_zeroline, t_zeroline])),
+                           coeff)
+    comp_zeroline[:, 1] = comp_zeroline[:, 1] + shift
+
+    comp1 = pca_fit['x1_fit']
+    c1_zeroline_prob = stats.invgauss.cdf(
+        comp_zeroline[:, 0], mu=comp1['mu'], loc=0, scale=comp1['scale'])
+
+    mu_slope = pca_fit['mu_fit'].slope
+    mu_intercept = pca_fit['mu_fit'].intercept
+    mu_zeroline = mu_slope * comp_zeroline[:, 0] + mu_intercept
+
+    sigma_polynomial_coeffcients = pca_fit['sigma_fit'].x
+    sigma_zeroline = np.polyval(
+        sigma_polynomial_coeffcients, comp_zeroline[:, 0])
+    c2_zeroline_prob = stats.norm.cdf(comp_zeroline[:, 1],
+                                      loc=mu_zeroline, scale=sigma_zeroline)
+
+    c1_normzeroline = stats.norm.ppf(c1_zeroline_prob, 0, 1)
+    c2_normzeroline = stats.norm.ppf(c2_zeroline_prob, 0, 1)
+
+    return_periods = np.asarray(return_periods)
+    contour_probs = 1 / (365*24*60*60/sea_state_duration * return_periods)
+
+    # Reliability contour generation
+    # Calculate reliability
+    beta_lines = stats.norm.ppf((1 - contour_probs), 0, 1)
+    # Add zero as lower bound to first contour
+    beta_lines = np.hstack((0, beta_lines))
+    # Discretize the circle
+    theta_lines = np.linspace(0, 2 * np.pi, 1000)
+    # Add probablity of 1 to the reliability set, corresponding to
+    # probability of the center point of the normal space
+    contour_probs = np.hstack((1, contour_probs))
+
+    # Vary U1,U2 along circle sqrt(U1^2+U2^2) = beta
+    u1_lines = np.dot(np.cos(theta_lines[:, None]), beta_lines[None, :])
+
+    # Removing values on the H_s = 0 line that are far from the circles in the
+    # normal space that will be evaluated to speed up calculations
+    minval = np.amin(u1_lines) - 0.5
+    mask = c1_normzeroline > minval
+    c1_normzeroline = c1_normzeroline[mask]
+    c2_normzeroline = c2_normzeroline[mask]
+
+    # Transform to polar coordinates
+    theta_zeroline = np.arctan2(c2_normzeroline, c1_normzeroline)
+    rho_zeroline = np.sqrt(c1_normzeroline**2 + c2_normzeroline**2)
+    theta_zeroline[theta_zeroline < 0] = theta_zeroline[
+        theta_zeroline < 0] + 2 * np.pi
+
+    sample_alpha, sample_beta, weight_points = _generate_sample_data(
+        beta_lines, rho_zeroline, theta_zeroline, points_per_interval,
+        contour_probs)
+
+    # Sample transformation to principal component space
+    sample_u1 = sample_beta * np.cos(sample_alpha)
+    sample_u2 = sample_beta * np.sin(sample_alpha)
+
+
```

Comment:
Double spaces are reserved for use between functions

Reply:
Done w/ @hivanov-nrel latest commit.

Reply:
Done w/ @hivanov-nrel latest commit.